Use Auth0 hosted login page and enable SSO

xoen opened the pull request:
### What
- Main aim of this is to start using Auth0 hosted login page instead of custom login page.
- I also enabled Silent SSO so that user is not bothered to login when they're already logged in central SSO.
- Updated `/logout` endpoint to log user out also from central SSO (relies on new environment variable `APP_HOST`)

### Why
Because Auth0 Lock.js < 11 are deprecated and will stop working on 1st April.

### Other changes
I also tidied up few things as I was in here:
- Expire session cookies after 1 hour
- Added proper `/healthz` healtcheck endpoint
- Switched to `passport-auth0-openidconnect` (from `passport-auth0`)
- Bumped all dependencies
- Removed all templates and assets and related dependencies (`pug`, `uglify-js`)
- Removed more unused dependencies:
  - `cookie-parser`, not required since `express-session` 1.5.0
  - `body-parser`, not used as we don't touch the requests body
  - `serve-favicon`, not used as we proxy these requests
- Started to use `dotenv` (was already a dependency but not used)
- Updated `Dockerfile to use Node.js 8.9.4 (LTS)
- Moved configuration to `config.js`
- Improved project structure a bit

### Ticket
https://trello.com/c/robNgbzG/651-3-update-kibana-auth-proxy-to-use-hosted-auth0-login

xoen committed:
Git-ignore .env file

xoen committed:
More sensible structure: App code under /app

xoen committed:
Stopped proliferation of config scattered around

Moved all configuration in app/config.js

NOTE: Not sure how/if the `SHINY_*` environment variables are used
      Probably not used as helm chart doesn't seem to pass them in!

xoen committed:
Expire session cookies after 1 hour (default)

Can be configured by passing the number of seconds via the `COOKIE_MAXAGE`
environment variable.

See: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control

xoen committed:
Removed views, public/static and pug dependency

As /login will start to use Auth0' Hosted login page.

xoen committed:
Removed unnecessary dependencies to unused packages

Removed:
- body-parser, it's a proxy, we don't mess with the body
- serve-favicon, was required but not used
- uglify-js, not used and we don't have templates anymore anyway

xoen committed:
Use Auth0 hosted login page and silent SSO (not working)

...and bumped dependencies versions

xoen committed:
Bumped Node version to 8.9.4 (LTS)

xoen committed:
Removed unnecessary dependency to cookie-parser module

express-session 1.5.0+ reads/write cookies directly:

From https://github.com/expressjs/session :

> Note Since version 1.5.0, the cookie-parser middleware no longer
> needs to be used for this module to work. This module now
> directly reads and writes cookies on req/res. Using cookie-parser
> may result in issues if the secret is not the same between this
> module and cookie-parser.

xoen committed:
Fixed login problems caused by misconfiguration

The `config.auth0.passReqToCallback` setting needs to be `true` or
the request will not be passed to the strategy and the session cookie
couldn't not be read.

xoen committed:
Don't store id_token in session (not necessary)

xoen committed:
Updated authorisation logic to work with OIDC complaiant client

We used a non-standard claim (`provided`) which is not included anymore
when the Auth0 client is OIDC complaiant.

As the provider is also the first part of the Auth0 user ID (`sub` is something
like `provider|ID`) we're extracting it from there.

andyhd commented on app/app.js:
```diff
@@ -0,0 +1,68 @@
+var config = require('./config');
+var routes = require('./routes');
+
+// This will configure Passport to use Auth0
```
Comments should explain _why_, not _what_

andyhd replied:
This one seems unnecessary

andyhd commented on app/app.js:
```diff
@@ -0,0 +1,68 @@
+Auth0Strategy.prototype.authorizationParams = (options) => options;
+passport.use(strategy);
+
+// you can use this section to keep a smaller payload
```
This comment is not very clear and not really necessary

andyhd commented on app/app.js:
```diff
@@ -0,0 +1,68 @@
+// This will configure Passport to use Auth0
+const strategy = new Auth0Strategy(
+  config.auth0,
+  ((req, issuer, audience, profile, accessToken, refreshToken, params, callback) => {
```
As the request is not used in this function, you should probably change the `passReqToCallback` setting to `false` in `config.auth0` and use the simpler function signature:
```javascript
(issuer, audience, profile, callback) => {
```

xoen replied:
I think that was the weird problem we had if you remember.
Even when not setting anything in the session explicitly here that `passReqToCallback` seems to be necessary - I'll give it another quick try but we tried last time and it wan't working without that set to `true`.

My theory was that if request is not passed back will not have cookie and therefore will not create user session - I don't know express enough to tell to be honest.

xoen replied:
I may be wrong but I think really depends on the strategy library used as I've notice the other one used in another project had a different callback signature.

Not sure if just rename and cut some of the params is safe here. I'll probably err on the safe side and leave the signature alone.

xoen replied:
FYI/To the future me: Tried again and yes, `passReqToCallback` needs to be `true` or it will not authenticate the user.

andyhd replied:
That function signature is from the [`passport-auth0-openidconnect` README](https://github.com/auth0/passport-auth0-openidconnect/blob/master/README.md).

If you set `passReqToCallback = false` you have to also change the function signature, it won't work if you do one or the other.

xoen replied:
OK, I'll give another shot to the this combination `passReqToCallback = false` and different signature and see if that works 👍

xoen replied:
It worked :) https://github.com/ministryofjustice/analytics-platform-kibana-auth-proxy/pull/1/commits/287e52ab54cf943f1801e949bd01dcc78e890fb5

andyhd commented on app/app.js:
```diff
@@ -0,0 +1,68 @@
+var Auth0Strategy = require('passport-auth0-openidconnect').Strategy;
+
+var config = require('./config');
+var routes = require('./routes');
```
Should all these `var`s be `const`s?

andyhd commented on app/config.js:
```diff
@@ -0,0 +1,45 @@
+config.shiny = {
+  host: process.env.SHINY_HOST,
+  port: process.env.SHINY_PORT,
+}
```
Should this section be renamed? Shiny isn't involved with this proxy, is it?

xoen replied:
I saw this weirdness but tried to ignore as I was wrestling with that authentication problem. You're right, I think this is a relic coming from another auth proxy. It should be safe to remove...hopefully!

xoen replied:
Yes, see: #1 (comment).

I'll try to tidy up further - I *think* I should be able to safely get rid of the code and this configs.

andyhd commented on app/index.js:
```diff
@@ -33,10 +37,7 @@ server.on('listening', onListening);
-      port: process.env.SHINY_PORT
-    }
+  target: config.shiny,
 });
```
Does this make sense? Either `SHINY_HOST` and `SHINY_PORT` are named incorrectly (should they be `KIBANA_HOST` and `KIBANA_PORT`?), or is this websockets proxy not needed? Does Kibana use websockets? Presumably they don't work anyway, because of ELB?

andyhd replied:
Why are there 2 proxy servers (one here and one in `routes.js`)?

xoen replied:
Yes, see: https://github.com/ministryofjustice/analytics-platform-kibana-auth-proxy/pull/1#discussion_r174488885

I'll try to tidy up further.

xoen replied:
https://github.com/ministryofjustice/analytics-platform-kibana-auth-proxy/pull/1/commits/f43d8380c7845af9e43cc7e61987d13a1501539d

andyhd commented on app/routes.js:
```diff
@@ -0,0 +1,85 @@
+  proxy.web(req, res);
+});
+
+var kibanaAuthCredsFromRequest = function(req){
```
this should either be `const kibanaAuthCredsFromRequest` or `function kibanaAuthCredsFromRequest` (I think the second is better)

xoen replied:
Yeah, I'm not quite sure why someone would want to define a function like this. 

I guess it's useful assign small-ish functions to variables sometimes but in this case no idea why it was like this - I'll change 👍

xoen replied:
FYI: https://github.com/ministryofjustice/analytics-platform-kibana-auth-proxy/pull/1/commits/1ae73a530214d055897886366a48953cf09ab229

andyhd commented on app/routes.js:
```diff
@@ -0,0 +1,85 @@
+  if (provider === 'google-oauth2') {
+    return config.kibana.adminCreds;
+  } else {
+    return config.kibana.creds;
```
The `else` is redundant. This should be
```javascript
if (provider === ...) {
  return ...adminCreds;
}
return ...creds
```

xoen replied:
I could change if you want.

xoen replied:
FYI: https://github.com/ministryofjustice/analytics-platform-kibana-auth-proxy/pull/1/commits/1ae73a530214d055897886366a48953cf09ab229

andyhd reviewed (changes requested):
Lots of unnecessary comments. I'm assuming you're not changing all the `var`s to `const`s to avoid irrelevant changes in this PR.

xoen commented:
As discussed earlier, I mainly focused on the authentication bit with some relatively light polishing (as the structure wasn't great, etc...).

I'll have a look at these comments and change some of the existing stuff further as long as I'm confident I don't risk to introduce bugs 👍 (we may want to do further refactoring independently if we want/need to)

xoen committed:
More tidying up (following PR comments).

I'm still trying to not change too many things at once here and keep
the refatorings to the minimum but changed few more things:
- Use `const` as much as possible
- Removed unnecessary comments

xoen committed:
PR amendment: Removed unused websocket proxy (and config)

Kibana is not using websockets and this is definitely not working
as:
1. Config was named "shiny" but this is not the shiny proxy so
   probably was copied from somewhere else
2. helm chart doesn't pass in these `SHINY_*` environment variables
   so they would be blank and yet this kibana-auth-proxy is working
   anyway

xoen committed:
PR amendment: Minor style change to kibanaAuthCredsFromRequest() function

xoen committed:
PR amendment: change `passReqToCallback` to false

Weirdly `passport-auth0-openidconnect`' strategy uses a different
function signature depending if the `passReqToCallback` option
is `true` or `false`.

As we don't need to access the session in our case we can use the
simplified function signature:

```javascript
function(issuer, audience, profile, cb)
```

and set `passReqToCallback` to `false`.

See: https://github.com/auth0/passport-auth0-openidconnect/blob/master/README.md

xoen commented:
Cheers @andyhd
